modules/apps/object/object-dynamic-data-mapping-form-field-type/src/main/java/com/liferay/object/dynamic/data/mapping/form/field/type/internal/attachment/AttachmentDDMFormFieldTemplateContextContributor.java

```diff
@@ -44,11 +56,62 @@
 			"acceptedFileExtensions",
 			ddmFormField.getProperty("acceptedFileExtensions")
 		).put(
-			"maximumFileSize",
-			ddmFormField.getProperty("acceptedFileExtensions")
+			"fileSource", ddmFormField.getProperty("fileSource")
+		).put(
+			"maximumFileSize", ddmFormField.getProperty("maximumFileSize")
+		).put(
+			"objectEntryId", ddmFormField.getProperty("objectEntryId")
+		).put(
+			"uploadURL",
+			_getUploadURL(
+				ddmFormField,
+				ddmFormFieldRenderingContext.getHttpServletRequest())
 		).put(
 			"value", ddmFormFieldRenderingContext.getValue()
 		).build();
 	}
 
+	private String _getPortletId(DDMFormField ddmFormField) {
+		ObjectField objectField = _objectFieldLocalService.fetchObjectField(
+			GetterUtil.getLong(ddmFormField.getProperty("objectFieldId")));
+
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectField.getObjectDefinitionId());
+
+		return objectDefinition.getPortletId();
+	}
+
+	private String _getUploadURL(
+		DDMFormField ddmFormField, HttpServletRequest httpServletRequest) {
+
+		String uploadURL = GetterUtil.getString(
+			ddmFormField.getProperty("uploadURL"));
+
+		if (Validator.isNotNull(uploadURL)) {
```

Comment:
I added this condition considering the case where the field is evaluated because it is required, for example. In this case, the field will already have the `uploadURL` because of the `DDMFormTemplateContextProcessor -> getDDMFormField` and we can avoid generating it again.

> Is it possible to tamper the URL?

I don’t see how.

Reply:
If this url is being pre filled because the entire field definition comes from the request, then it's possible to tamper the url before it reaches the server. But now that I think about this, we won't be able to prevent this from being tampered because, we send this to the client after all and this could be altered at the response to the client. So let's not change this, but maybe we should think about validating the folderId and objectEntryId at the AttachmentValidator. Let me know what you think.